packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+#include "packager/VisibilityChecker.h"
+#include "ast/treemap/treemap.h"
+#include "common/concurrency/ConcurrentQueue.h"
+#include "common/formatting.h"
+#include "common/sort.h"
+#include "core/Context.h"
+#include "core/errors/packager.h"
+#include "core/errors/resolver.h"
+#include <iterator>
+
+namespace sorbet::packager {
+
+namespace {
+
+// For each __package.rb file, traverse the resolved tree and apply the visibility annotations to the symbols.
+class PropagateVisibility final {
+    const core::packages::PackageInfo &package;
+
+    bool definedByThisPackage(const core::GlobalState &gs, core::ClassOrModuleRef sym) {
+        auto pkg = gs.packageDB().getPackageNameForFile(sym.data(gs)->loc().file());
+        return this->package.mangledName() == pkg;
+    }
+
+    void recursiveExportSymbol(core::GlobalState &gs, bool firstSymbol, core::ClassOrModuleRef klass) {
+        // There is a loop between a class and its singleton, and this is the easiest way to detect it.
+        if (!firstSymbol && klass.data(gs)->flags.isExported) {
+            return;
+        }
+
+        // We only mark symbols from this package.
+        if (!this->definedByThisPackage(gs, klass)) {
+            return;
+        }
+
+        klass.data(gs)->flags.isExported = true;
+
+        for (const auto &child : klass.data(gs)->members()) {
+            if (child.second.isClassOrModule()) {
+                recursiveExportSymbol(gs, false, child.second.asClassOrModuleRef());
+            } else if (child.second.isFieldOrStaticField()) {
+                child.second.asFieldRef().data(gs)->flags.isExported = true;
+            }
+        }
+    }
+
+    void exportParentNamespace(core::GlobalState &gs, core::ClassOrModuleRef owner) {
+        while (owner.exists() && !owner.data(gs)->flags.isExported && this->definedByThisPackage(gs, owner)) {
+            owner.data(gs)->flags.isExported = true;
+            owner = owner.data(gs)->owner;
+        }
+    }
+
+    // Lookup the package name on the given root symbol, and mark the final symbol as exported.
+    void exportRoot(core::GlobalState &gs, core::ClassOrModuleRef sym) {
+        // Explicitly mark the symbol outside of the special package namespace as exported by this package, to allow
+        // other packages to refer to this namespace. Ultimately it would be great to not have this, but it helps with a
+        // migration to the visibility checker from the previous packages implementation.
```

Comment:
Sorry I don't quite understand what this comment is saying. Could you give more color here?

Reply:
For a package `A::B`, the symbol that we'll be processing in the context of the `__package.rb` file will be named `<PackageRegistry>::A::B`. Part of the current package implementation requires that you be able to use the name of a package if it's imported, so this captures that by explicitly exporting the constants `A::B` and `Test::A::B`. I'll rephrase this a bit.